pkg/controllers/provisioning/v1alpha1/reallocator/filter.go

```diff
@@ -0,0 +1,50 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reallocator
+
+import (
+	"context"
+	"fmt"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Filter struct {
+	kubeClient client.Client
+}
+
+// Gets Nodes fitting some underutilization
+// TODO: add predicate
+func (f *Filter) GetUnderUtilizedNodes(ctx context.Context) ([]*v1.Node, error) {
+	// 1. Get all nodes
+	nodeList := &v1.NodeList{}
+
+	// 2. Get nodes and the pods on each node
+	for _, node := range nodeList.Items {
```

Comment:
Might be cleaner to orient this with a predicate
```
for _, node := range ...
  pods, err := ...
  if err ...
  if isUnderutilized(node, pods) {
     underutilizedNodes = append(underutilizedNodes, node)
  }
```

This is actually generalizable to 
```
filtered := {}
for _, item := range items {
  if predicate(item) {
    filtered = append(filtered, item)
  }
}
```
You will encounter this pattern many times in your career :)

Reply:
Or in a functional language like javascript
```
items = items.filter(predicate)
```